lms/djangoapps/courseware/block_render.py

```diff
@@ -364,60 +372,132 @@ def display_access_messages(user, block, view, frag, context):  # pylint: disabl
-        user_location=user_location,
-        request_token=xblock_request_token(request),
-        disable_staff_debug_info=disable_staff_debug_info,
-        course=course,
-        will_recheck_access=will_recheck_access,
+    if request:
+        track_function = track_function or make_track_function(request)
+        user_location = user_location or getattr(request, 'session', {}).get('country_code')
+        request_token = request_token or xblock_request_token(request)
+
+    if not student_data:
+        student_kvs = DjangoKeyValueStore(field_data_cache)
+        if is_masquerading_as_specific_student(user, course_key):
+            student_kvs = MasqueradingKeyValueStore(student_kvs, request.session)
+        student_data = KvsFieldData(student_kvs)
+
+    # The runtime is already shared between XBlocks. If there are no wrappers, it is the first initialization.
+    should_recreate_runtime = not block.runtime.wrappers
+
+    # If the runtime was prepared for another user, it should be recreated.
+    # This part can be removed if we remove all user-specific handling from the runtime services and pull this
+    # information directly from XBlocks during the service initialization.
+    if not should_recreate_runtime:
+        # If the user service is absent (which should never happen), the runtime should be reinitialized.
+        # We retrieve this service directly to bypass service declaration checks.
+        if user_service := block.runtime._services.get('user'):  # pylint: disable=protected-access
+            # Check the user ID bound to the runtime. This operation can run often for complex course structures, so we
+            # are accessing the protected attribute of the user service to reduce the number of queries.
+            should_recreate_runtime = user.id != user_service._django_user.id  # pylint: disable=protected-access
+        else:
+            should_recreate_runtime = True
+
+    if should_recreate_runtime:
+        prepare_runtime_for_user(
+            user=user,
+            student_data=student_data,  # These have implicit user bindings, the rest of args are considered not to
+            runtime=block.runtime,
+            course_id=course_key,
+            track_function=track_function,
+            wrap_xblock_display=wrap_xblock_display,
+            grade_bucket_type=grade_bucket_type,
+            static_asset_path=static_asset_path,
+            user_location=user_location,
+            request_token=request_token,
+            disable_staff_debug_info=disable_staff_debug_info,
+            course=course,
+            will_recheck_access=will_recheck_access,
+        )
+
+    # Pass position specified in URL to runtime.
+    if position is not None:
+        try:
+            position = int(position)
+        except (ValueError, TypeError):
+            log.exception('Non-integer %r passed as position.', position)
+            position = None
+
+    block.runtime.set('position', position)
```

Comment:
As far as I understand, this is used by the `SequenceBlock`/subsection to determine which tab should be active:
https://github.com/openedx/edx-platform/blob/0b455e0336fa54199fd992c2ce83a81c07dd07c2/xmodule/seq_block.py#L335-L363

This is no longer used in the learning MFE, but the Preview still uses the old experience. I tried removing this, but it caused some errors. It should be safe to remove when we completely remove the legacy experience. For now, I would leave it as-is.

Reply:
I see. Could you perhaps add that information as a comment into the function?

Reply:
Sure, I added it in d7b898. I added this information to the `get_block` function since it contains the description of these parameters. Let me know if you had another function in mind.